app/controllers/papers_controller.rb

```diff
@@ -221,7 +221,18 @@ def show
     # Don't show the paper to anyone other than the submitting author or an
     # admin.
     if @paper.invisible?
-      head 404 and return unless can_see_hidden_paper?(@paper)
+      # Redirect to login if not logged in
+      if !current_user
+        flash[:notice] = "You need to log in before viewing this paper."
```

Comment:
@xuanxu – I think we can just bounce people to the front page if they're not logged in. (IMO this is preferable to the current behaviour)